[Sema] Diagnose internal(set) from @inlinable functions

harlanhaskins opened the pull request:
This patch mainly consolidates the functions used to check accessors vs.
other decls, and makes sure we check setter access as well as regular
decl access.

rdar://45217648
Resolves [SR-8969](https://bugs.swift.org/browse/SR-8969)

harlanhaskins commented:
@swift-ci please test

jrose-apple commented on lib/Sema/ResilienceDiagnostics.cpp:
```diff
@@ -138,17 +138,24 @@ bool TypeChecker::diagnoseInlinableDeclRef(SourceLoc loc,
 
+  // For accessors, use the name of the storage instead of the implicit '_'.
+  auto accessor = dyn_cast<AccessorDecl>(D);
+  auto printedName = accessor ?
```
Nitpick: this isn't really a "printed" name, since it's a DeclName.

harlanhaskins commented:
@swift-ci please test

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -4104,27 +4104,31 @@ ERROR(usable_from_inline_attr_in_protocol,none,
 NOTE(resilience_decl_declared_here,
-     none, "%0 %1 is not '@usableFromInline' or public", (DescriptiveDeclKind, DeclName))
+     none, DECL_OR_ACCESSOR "2 %1 is not '@usableFromInline' or public",
+     (DescriptiveDeclKind, DeclName, bool))
```
Can you make sure each of these diagnostics is tested?

harlanhaskins replied:
There currently aren't any tests for the `downgradeToWarning` variant of this. Should I copy `test/attr/attr_inlinable.swift` and rewrite the `expected-error`s?

jrose-apple replied:
Yes please! cc @slavapestov just to make sure they're not hidden somewhere.

slavapestov replied:
Take a look at `test/Compatibility/attr_inlinable*.swift`

jrose-apple commented:
Note that if we can't take this on the Swift 5 branch it all gets trickier w/r/t source compat / when to use warnings vs. errors…

harlanhaskins commented:
@swift-ci please test

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -2439,8 +2439,7 @@ bool ValueDecl::isUsableFromInline() const {
       return true;
 
   if (auto *containingProto = dyn_cast<ProtocolDecl>(getDeclContext())) {
-    if (isProtocolRequirement() &&
```
This change should be a separate commit with its own test case:

```
@usableFromInline protocol P {
  typealias T = Int
}

extension P {
  @inlinable func f() {
    _ = T.self
  }
}
```

slavapestov reviewed (approved):
Looks good except for one nit

jrose-apple commented on test/attr/attr_inlinable.swift:
```diff
@@ -248,3 +248,11 @@ public struct KeypathStruct {
+  public internal(set) var x: Int // expected-note {{setter for 'x' is not '@usableFromInline' or public}}
+
+  @inlinable public mutating func setsX() {
+    x = 10 // expected-error {{setter for 'x' is internal and cannot be referenced from an '@inlinable' function}}
```
Oops, wait, we can't make this an error in 4.2 mode.

jrose-apple replied:
@slavapestov Any preference? I think we've been leaning towards always putting the latest behavior in the "normal" file and older behavior in "Compatibility".

harlanhaskins replied:
Went ahead and just did that

harlanhaskins committed:
[test] Update @inlinable test to Swift 5

Also add a compatibility test for Swift 4.2 behavior

harlanhaskins committed:
[Sema] Non-requirement protocol members should inherit @usableFromInline

Previously, members of protocols that were not protocol requirements,
like accessors and typealiases, did not inherit @usableFromInline from
the parent protocol. Change this so they do.

harlanhaskins committed:
[Sema] Diagnose internal(set) from @inlinable functions

This patch mainly consolidates the functions used to check accessors vs.
other decls, and makes sure we check setter access as well as regular
decl access.

rdar://45217648

harlanhaskins commented:
@swift-ci please test
